### PR TITLE
Implements visitor pattern

### PR DESCRIPTION
**Pull request by koriym**
Implementing the visitor pattern in Ray.Di for a full rewrite of Ray.Compiler.

image

See https://speakerdeck.com/koriym/daburudeisupatutitobizitapatan

## Summary by CodeRabbit

- **New Features**
  - Introduced an `AcceptInterface` across multiple classes to enable visitor pattern functionality.
  - Added a `NullVisitor` class to handle visits to various dependency injection components.

- **Bug Fixes**
  - Updated PHP versions in CI workflow for improved stability.

- **Documentation**
  - Marked `NullCache` class as deprecated.

- **Tests**
  - Added `testSingleArray()` method in `InjectorTest` for enhanced testing coverage.

### CODE REVIEW

**Comment by koriym** [comment]
@coderabbitai review

**Comment by koriym** [comment]
@coderabbitai review

**Comment by koriym** [comment]
@coderabbitai approve?

**Comment by koriym** [comment]
@coderabbitai I wanted to organise and strengthen this library, which has a 10-year history, to take it into the next decade.

**Comment by koriym** [comment]
@coderabbitai 先ほどの回答を日本語でもお願いします。

File: composer.json [hunk 1]
```diff
@@ -11,12 +11,9 @@
     ],
     "require": {
         "php": "^7.2 || ^8.0",
-        "doctrine/cache": "^1.10 || ^2.1",
```
Comment (NaokiTsuchiya) [inline composer.json:14]: もともとコンパイラーが一緒だったから doctrine/cache や php-parser が含まれているんでしたっけ？

**Review by NaokiTsuchiya — APPROVED** [review]

File: composer.json (hunk 1 above)
Comment (koriym) [reply composer.json:14]: はい、その名残だったと思います。 #288 も作ってました。

**Pull request merged**